src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+                                         predictedInsertSize: Option[Integer] = None,
+                                         runDate: Option[Iso8601Date] = None,
+                                         platformModel: Option[String] = None,
+                                         programGroup: Option[String] = None,
+                                         comments: List[String] = Nil
+                                        ): SAMFileHeader = {
+    val header: SAMFileHeader = new SAMFileHeader
+    val libraryId: String     = sample.libraryId.getOrElse(sample.sampleName)
+    val platformUnit: String  = sample.sampleBarcode.map(sb => s"${sample.sampleName}.$sb").getOrElse(sample.sampleName)
+    val readGroup             = new SAMReadGroupRecord(sample.sampleId)
+    readGroup.setSample(sample.sampleName)
+    readGroup.setLibrary(libraryId)
+    readGroup.setPlatformUnit(platformUnit)
+    sample.description.foreach(readGroup.setDescription)
+    platform.foreach(readGroup.setPlatform)
+    sequencingCenter.foreach(readGroup.setSequencingCenter)
+    predictedInsertSize.foreach(readGroup.setPredictedMedianInsertSize)
+    runDate.foreach(readGroup.setRunDate)
+    platformModel.foreach(readGroup.setPlatformModel)
+    programGroup.foreach(readGroup.setProgramGroup)
+    header.addReadGroup(readGroup)
+    comments.foreach(header.addComment)
+    header.setSortOrder(sortOrder)
+    header
+  }
+
+  /** Creates SAM file writers for each sample in the Sample Sheet, returning them in the same order as specified
+    * by each sample's ordinal.  An extra writer is appended for records that do not match a sample barcode. */
+  private def createSamFileWriters(sampleSheet: SampleSheet,
+                                   output: DirPath,
+                                   unmatched: String,
+                                   sampleToHeader: Sample => SAMFileHeader): List[SAMFileWriter] = {
+    def toWriter(header: SAMFileHeader, outputBam: PathToBam): SAMFileWriter = {
+      new SAMFileWriterFactory()
+        .setUseAsyncIo(UseAsyncIo)
+        .setMaxRecordsInRam(MaxRecordsInRamPerSamFileWriter)
+        .makeSAMOrBAMWriter(header, header.getSortOrder == SortOrder.unsorted, outputBam.toFile)
+    }
+
+    val writers: List[SAMFileWriter] = sampleSheet.map { sample =>
+      toWriter(sampleToHeader(sample), sampleOutputBam(output, sample))
+    }.toList
+    val unmatchedWriter = {
+      val outputBam: FilePath   = output.resolve(unmatched)
+      val header: SAMFileHeader = sampleToHeader(new Sample(writers.length, UnmatchedSampleId, UnmatchedSampleId, Some(UnmatchedSampleId), None, None))
+      toWriter(header, outputBam)
+    }
+    writers ++ List(unmatchedWriter)
+  }
+
+  /** Gets the quality format of the FASTQs. */
+  private def determineQualityFormat(fastqs: Seq[PathToFastq], qualityFormat: Option[FastqQualityFormat] = None, logger: Option[Logger] = None): FastqQualityFormat = {
```

Comment:
It's actually the expected quality format.  If it's `None`, it guesses the best format, and if it is `Some(format)` then it verifies the quality format.  This is the same behavior as Picard's `FastqToSam`, which I am trying to keep a little bit of.